### ISSUE TITLE
Support for emotion

### ISSUE THREAD

**Issue by zhaoyao91** [Wildhoney/ReactShadow#73]
look through the code, I found it have some part dedicated to handle styled-component.

is it ok to support emotion too?

BTW, I think it could be better do not require these packages since it  troubling the bundler.

**Comment by Wildhoney** [comment]
Does Emotion support the `StyleSheetManager` like Styled Components does? Where you can tell it where to renders parts of the styles.

I have thought about how to solve the issue of the `require` &mdash; I can't `async import` because it needs to be a sync import, and I can't place the `import` at the top because many projects won't have `styled-components` installed, nor do I want to bundle it.

At that point I'm really only left with creating a separate file, so people would have to `import react-shadow/styled-components` which may not be too bad of a solution. Thoughts?

**Comment by zhaoyao91** [comment]
Emotion has [CacheProvider](https://emotion.sh/docs/cache-provider), which works similarly.

I don't think the styled-component or emotion binding should be external to this repo as they are outside extensions and as a base repo you can't expect what the worlds has or will have.

I have created an emotion binding since this issue, see https://github.com/zhaoyao91/react-emotion-shadow

**Comment by Wildhoney** [comment]
Yep, but having the files as part of the repo would work just fine.

- `import react-shadow` would give you the base;
- `import react-shadow/styled-components` would give you the base + `styled-components`;
- `import react-shadow/emotion` would give you the base + `emotion`;

Obviously the 2 optional dependencies would not be bundled as part of the project. That would remove the need for `require` as the deps could be imported using standard `import`. Only if you imported `react-shadow/emotion` and didn't have Emotion installed as part of your project would you run into issues.

**Comment by zhaoyao91** [comment]
> import react-shadow would give you the base;
import react-shadow/styled-components would give you the base + styled-components;
import react-shadow/emotion would give you the base + emotion;

@Wildhoney this is workable, as you can maintain the list of popular solutions :)

**Comment by Wildhoney** [comment]
All right 👍 I'll get onto this a little later.

Thanks for your input @zhaoyao91 &mdash; I might need some help with the Emotion implementation, but your example helps too.

**Comment by Wildhoney** [comment]
Could you take a glance over the Emotion [implementation](https://github.com/Wildhoney/ReactShadow/blob/master/src/emotion/index.js) and [example](https://github.com/Wildhoney/ReactShadow/blob/master/example/components/Header/index.js) usage please @zhaoyao91?

**Comment by zhaoyao91** [comment]
I have had a try, it worked well in your example, I am considering to deprecate my package :)

![image](https://user-images.githubusercontent.com/3808838/78124782-c03b5980-7442-11ea-8b65-a7cfd24d13af.png)

there may be a useMemo.

and, is there a chance to pass options to the createCache call?

**Comment by Wildhoney** [comment]
I'll have a think on the best way to support passing options to `createCache`. I've published `v18.0.0` which adds the Emotion support, so you can give it a try 🤞.  

Does `emotion` support SSR where you can collect the `style` tags of `children` and render them to a HTML string?

With `styled-components` you can do:

```javascript
const sheet = new ServerStyleSheet();

renderToString(
    <StyleSheetManager sheet={sheet.instance}>
        {children}
    </StyleSheetManager>,
);

return sheet.getStyleElement();
```

**Comment by zhaoyao91** [comment]
emotion does support ssr, see https://emotion.sh/docs/ssr
but I am not experienced at ssr, not able to give helpful comment. you may have a look at the link above💪
